backend/src/auth.js

```diff
@@ -18,77 +15,37 @@ function initAuth(app) {
   app.use(passport.session());
 
   // For saving user creds in cookie
-  passport.serializeUser((userCreds, done) => {
-    return done(null, userCreds.id);
-  });
+  passport.serializeUser(({ id }, done) => done(null, id));
 
   // For retrieving user creds object from cookie
-  passport.deserializeUser((userCredsIdString, done) => {
-    const userCredsId = mongoose.Types.ObjectId(userCredsIdString);
-    UserCreds.findById(userCredsId, (err, userCreds) => {
-      return done(err, userCreds);
-    });
-  });
+  passport.deserializeUser((id, done) => UserData.findById(id, done));
 
   // Google Auth config via passport
   passport.use(
-    new GoogleStrategy(
+    new GoogleTokenStrategy(
       {
         clientID: process.env.GOOGLE_CLIENT_ID,
-        clientSecret: process.env.GOOGLE_CLIENT_SECRET,
-        callbackURL: PUBLIC_FACING_AUTH_CALLBACK_URL
+        clientSecret: process.env.GOOGLE_CLIENT_SECRET
       },
       (accessToken, refreshToken, profile, done) => {
-        UserCreds.findOne({ googleId: profile.id })
-          .then(userCreds => {
-            if (userCreds) {
-              // User credentials were found, first update access and refresh tokens
-              userCreds.accessToken = accessToken;
-              userCreds.refreshToken = refreshToken;
-              return userCreds.save();
-            }
-
-            // First time login, create the UserCreds document
-            const newUserCreds = new UserCreds({
-              googleId: profile.id,
-              accessToken,
-              refreshToken,
-              userDataId: null // User does not have any data in the system yet
-            });
-            return newUserCreds.save();
-          })
-          .then(savedUserCreds => {
-            // Return updated user creds
-            done(null, savedUserCreds);
-          })
-          .catch(err => {
-            done(err);
+        UserCreds.upsertGoogleUser(accessToken, refreshToken, profile, function(err, userCred) {
+          if (err) return done(err, null);
+          UserData.findById(userCred.userDataId, (err2, user) => {
+            return done(err2, user);
```

Comment:
I'm not 100% sure, but I believe `user` can be null if the ID doesn't exist. Do we have to handle that case separately, or does `done` accept a null `user`?

Reply:
So I don't think that situation will happen because inside `UserCreds.upsertGoogleUser()` it creates a UserData object for a new UserCreds object. Only way this may come up is if a UserData object gets deleted and the corresponding UserCreds object is not deleted? Is that a case we may encounter?

Also, I'm not sure how we would even handle that case with the `done()` function. I believe if we return `done(null, null)` it just means the user won't be saved in `req.user`? Which with our auth checks will be treated as an invalid login?

Reply:
Good point, this could happen if we accidentally delete data in the DB, but shouldn't occur in any normal case. Plus `done` probably will take care of null as you mentioned, so this is OK.